### PR TITLE
New package: code-server

### PR DESCRIPTION
**Pull request by qwertos**

### CODE REVIEW

**Comment by adamjstewart** [comment]
Is it possible to install this package from source? The binaries distributed won't be optimized for any hardware, and won't support any other OS.

**Comment by qwertos** [comment]
I looked into doing a source build, and the devs provide zero documentation for that. They seem to be in the mindset of "why would anyone want to build from source when we have this beautiful docker image and binary build". I'll take another look at it.

**Comment by hartzell** [comment]
> "why would anyone want to build from source when we have this beautiful docker image and binary build"

We need to do a [Demotivational Poster](https://despair.com/collections/demotivators), "Docker, building from source is a thing of the past!".

So sad!

**Comment by adamjstewart** [comment]
@qwertos is this still a work in progress?

**Comment by qwertos** [comment]
Technically, yes. Functionally, it's so far down on my to do list that I don't know when I'm going to get back to it. I won't object to someone picking up the torch (or merging as is using the pre-built binaries).

File: var/spack/repos/builtin/packages/code-server/package.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    version('3.1.0', sha256='5ef85c8f280ce781a176a8b77386b333efe892755a5c325a1782e4eac6016e59')
+
+    depends_on('git@2:')
+
+    def install(self, spec, prefix):
+        ln = which('ln')
+        mkdir = which('mkdir')
+        cp = which('cp')
+
+        cp('-r', '{0}/.'.format(self.stage.source_path), prefix)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/code-server/package.py]:
```suggestion
        install_tree('.', prefix)
```

File: var/spack/repos/builtin/packages/code-server/package.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        mkdir(prefix.bin)
+        ln('-s', '{0}/code-server'.format(prefix), prefix.bin)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/code-server/package.py]:
```suggestion
        symlink('{0}/code-server'.format(prefix), prefix.bin)
```

File: var/spack/repos/builtin/packages/code-server/package.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+    version('3.1.1', sha256='5dd922d28b2e351c146081849d987fb1e439ee7d53b941434b2eecb2a194da71')
+    version('3.1.0', sha256='5ef85c8f280ce781a176a8b77386b333efe892755a5c325a1782e4eac6016e59')
+
+    depends_on('git@2:')
+
+    def install(self, spec, prefix):
+        ln = which('ln')
+        mkdir = which('mkdir')
+        cp = which('cp')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/code-server/package.py]: You can remove these. Spack has builtin tools for all of these that will work regardless of OS. We plan on adding Windows support in the near future.

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Let's just merge as is with the binary package and someone can figure out the source build later. Can you make a few changes to the install method?

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**